openassessment/xblock/job_sample_grader/code_grader.py

```diff
@@ -152,47 +154,89 @@ def run_code(self, run_type, executor_id, source_code, problem_name):
                     'error': None,
                 }
         """
-        test_case_paths = glob.glob(
-            '{}{}/{}/*'.format(self.__SECRET_DATA_DIR__, problem_name, run_type)
-        )
-
-        # Sort the test cases based on the test number
-        if test_case_paths:
-            test_case_paths = sorted(
-                test_case_paths, key=lambda test_case: int(test_case.split('/')[-1])
-            )
+        xblock_id = self.get_xblock_id()
+        usage_key = xblock_id.split('@')[-1]
 
         test_case_files = []
-        for case in test_case_paths:
-            case_number = int(case.split('/')[-1])
-            input_file = '{}/input.in'.format(case)
-            expected_output_file = '{}/output.out'.format(case)
+        question_mapping = AssessmentQuestionXblockMapping.objects.filter(usage_key=usage_key).first()
 
-            # TODO: We can read cases from db here.
-            with open(input_file, 'rb') as file:
-                input_content = file.read()
+        if question_mapping:
+            # Read test cases from DB
+            question = question_mapping.question
 
-            test_case_files.append(
-                {
-                    'case_number': case_number,
-                    'input_file': {
-                        # Keeping the file names the same as host.
-                        # This will allow us to use the same names for epicbox and server_shell.
-                        'name': input_file,
-                        'content': input_content,
-                    },
-                    'expected_output_file': {'name': expected_output_file},
-                }
+            try:
+                test_cases = json.loads(question.metadata)[run_type]
+            except:
+                test_cases = question.metadata[run_type]
+
+            for case_number in test_cases:
+                case = test_cases[case_number]
+                input_file = NamedTemporaryFile(
+                    mode="w+",
+                    dir=self.__SECRET_DATA_DIR__,
+                    delete=False
+                )
+                input_file.write(case['input'])
+                input_file.seek(0, 0)
+
+                expected_output_file = NamedTemporaryFile(
+                    mode="w+",
+                    dir=self.__SECRET_DATA_DIR__,
+                    delete=False
+                )
+                expected_output_file.write(case['output'])
+                expected_output_file.seek(0, 0)
+
+                test_case_files.append(
+                    {
+                        'case_number': case_number,
+                        'input_file': {
+                            'name': input_file.name,
+                            'content': bytes(case['input'], 'utf-8'),
+                        },
+                        'expected_output_file': {'name': expected_output_file.name},
+                    }
+                )
+        else:
```

Comment:
Let's refactor this code. This function is huge.

Can we do something like self.get_test_case_files() ?